packages/records/lib/models/records.ts

```diff
@@ -216,53 +216,90 @@ export async function upsert({
-                            const errorCode = (res as { code: string }).code;
-                            return errorCode === '40P01'; // deadlock_detected
+                const withRetry = async <T>(query: Knex.QueryBuilder<any, T>) => {
+                    return retry(() => query, {
+                        maxAttempts: 3,
+                        delayMs: 500,
+                        retryIf: (res) => {
+                            if ('code' in res) {
+                                const errorCode = (res as { code: string }).code;
+                                return errorCode === '40P01'; // deadlock_detected
+                            }
+                            return false;
                         }
-                        return false;
-                    }
-                });
+                    });
+                };
 
-                const delta = chunkSummary.addedKeys.length - (chunkSummary.deletedKeys?.length ?? 0);
-                if (delta !== 0) {
-                    await trx
-                        .from(RECORD_COUNTS_TABLE)
-                        .insert({
-                            connection_id: connectionId,
-                            model,
-                            environment_id: environmentId,
-                            count: delta
-                        })
-                        .onConflict(['connection_id', 'environment_id', 'model'])
-                        .merge({
-                            count: trx.raw(`${RECORD_COUNTS_TABLE}.count + EXCLUDED.count`),
-                            updated_at: trx.fn.now()
-                        });
+                // We split insertions and updates into two queries
+                // in order to be able to know which records were inserted and which were updated
+
+                // we first insert new records and ignore conflicts
+                const insertQuery = trx
+                    .with('insert', (qb) => {
+                        qb.insert(encryptedRecords)
+                            .into(RECORDS_TABLE)
+                            .onConflict(['connection_id', 'external_id', 'model'])
+                            .ignore()
+                            .returning<{ external_id: string }[]>('external_id');
+                    })
+                    .from('insert')
+                    .pluck<string[]>('external_id');
+                const inserted = await withRetry(insertQuery);
+
+                // then update records that were not inserted
+                const updateQuery = trx
+                    .with('update', (qb) => {
+                        qb.insert(encryptedRecords)
+                            .into(RECORDS_TABLE)
+                            .onConflict(['connection_id', 'external_id', 'model'])
+                            .merge()
+                            .whereNotIn(`${RECORDS_TABLE}.external_id`, inserted)
+                            .returning('external_id');
```

Comment:
the upsert is now 2 queries but the initial query to determine the counts is not there anymore